In [1]:
import proj_functions
from importlib import reload
reload(proj_functions)
from proj_functions import *

# Problem 1

**(a)** Please check the function `factorize_with_primes(n,ps)` in proj_functions.py

**(b)** Please check the function `pseudoprime_test(n, num_trial)` in proj_functions.py

**(c)** Run the pseudoprime test for 3004879 with trials value of [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41]. Witness 2 is found. 3004879 is not a prime.

In [2]:
trials = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41]
proba_prime = pseudoprime_test(3004879,trials)

Target: 3004879 Pseudo Test Result: witness_found: 2, Not prime


# Problem 2

**(a)** Please check the function `jacobi_symbol` in proj_functions.py

**(b)** Please check the function `extended_eclid` in proj_functions.py

**(c)** Using extended Euclidean method, we found that $1163 \equiv 648^{-1} \pmod{1999}$.

In [3]:
_, a, _ = extended_euclid(1163,1999)
print(f'The multiplicative inverse of 1163 modulo 1999 is {a%1999}')

The multiplicative inverse of 1163 modulo 1999 is 648


**(d)** Please check the function `tonelli(a,p)` in proj_functions.py

**(e)** Use Tonelli function, we found one square root $2188^2 \equiv 117 \pmod{3691}$. The $i_k$ we get is [0,2] because $3691 = 2^1 \times 1845$. Thus there is no updates after the first two $i_k$.

In [4]:
root, ik = tonelli(117,3691, verbose = 1)
print(f'i_k = {ik}')

Tonelli Algorithm for square root of 117 modulo 3691...Sqrt Root found: 2188
i_k = [0, 2]


# Problem 3

**(a)** Please check the function `gauss_modulo2` in proj_functions.py

**(b)** Please check the function `find_dependent_rows` in proj_functions.py

**(c)** Create a helper function `read_file` to read data. The final dependency coefficients $a_r$ are given below. The value is already modulus 2,

In [5]:
M = transpose(read_file('sample.txt'))
coefs = left_null_space_mod2(transpose(M))
print(f'a_1 mod 2 = {coefs[0]}\n')
print(f'a_2 mod 2 = {coefs[1]}')

a_1 mod 2 = [0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0]

a_2 mod 2 = [1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1]


# Problem 4

**(a)** Please check the function `list_primes(B,n = 1)` in proj_functions.py

**(b)** Please check the function `list_primes(B,n = 1)` in proj_functions.py

**(c)** Use half sieve of eratosthenes, all primes less than 200 with $(91/p)=1$ is listed below. The function only works for small $B$

In [6]:
print(f'Primes p below 200 and (91/p) =1: {list_primes(200,n=91)}')

Primes p below 200 and (91/p) =1: [2, 3, 5, 11, 29, 41, 53, 67, 71, 73, 89, 97, 103, 113, 131, 139, 151, 163, 199]


# Problem 5

**(a)** Please check the funciton `calcualte_h(n,M)` in proj_functions.py

**(b)** Please check the funciton `generate_factor_base(n,B)` in proj_functions.py

**(c)** with $n = 7386829$ and $B = 1000$. We have the list of candidate prims and the factor base. Although there is no $t$ for $p=2$. However, 2 is still in the factor base.

In [7]:
ts,ps = solve_factor_base(7386829, 1000)
print('Factor base and Tonelli solution pairs {p: t}:')
print(ts)
print(f'With total {len(ps)} factors, 2 is also in Factor base but have no t')

Factor base and Tonelli solution pairs {p: t}:
{3: 2, 5: 2, 7: 3, 19: 3, 29: 6, 31: 26, 37: 36, 41: 8, 43: 26, 47: 11, 53: 31, 59: 18, 61: 41, 71: 42, 73: 18, 79: 48, 83: 24, 97: 45, 103: 94, 127: 126, 131: 130, 137: 101, 139: 62, 149: 68, 163: 141, 167: 69, 179: 6, 181: 20, 223: 20, 233: 27, 257: 69, 269: 122, 277: 65, 283: 260, 293: 251, 307: 193, 311: 184, 313: 177, 317: 290, 331: 137, 337: 183, 353: 294, 359: 274, 367: 68, 373: 181, 379: 285, 401: 295, 409: 392, 457: 126, 491: 476, 499: 272, 541: 189, 557: 190, 563: 371, 569: 54, 571: 85, 577: 464, 587: 487, 613: 479, 617: 71, 641: 540, 661: 223, 683: 54, 691: 102, 733: 467, 743: 573, 751: 155, 787: 566, 797: 673, 821: 604, 827: 723, 853: 629, 863: 570, 881: 273, 911: 21, 941: 866, 953: 850, 971: 648, 977: 705, 983: 741, 991: 887, 997: 71}
With total 83 factors, 2 is also in Factor base but have no t


**(d)** Please check the function `run_sieve(n,M,B)` in proj_function.py

**(e)** Please check the function `run_sieve(n,M,B)` in proj_function.py

**(f)** Here is the running with same factor base. The resulting matrix is $113\times83$ with more row than columns which suggest the Gaussian elimination is valid.

In [8]:
mat, r, _ = run_sieve(7386829, 2000, 1000)
print(f'M = 2000, Size of matrix: {len(mat)}x{len(mat[0])}')
print(f'Candidate r (or t): {r}')

M = 2000, Size of matrix: 113x83
Candidate r (or t): [204, 936, 2592, 2754, 2756, 2770, 2772, 2778, 2790, 2792, 2796, 2826, 2842, 2846, 2850, 2868, 2878, 2880, 2882, 2888, 2908, 1260, 2940, 2950, 2962, 2972, 2978, 2996, 3012, 3014, 3050, 108, 3070, 3104, 3112, 3126, 3142, 3156, 3176, 3184, 3188, 3238, 3242, 3266, 3274, 3294, 3312, 3322, 3366, 3422, 3428, 3436, 3440, 3468, 3508, 3518, 3522, 3536, 3538, 3544, 3558, 3570, 3588, 3600, 3636, 3664, 3682, 3738, 3778, 3888, 3890, 3912, 3924, 3932, 3934, 3942, 3944, 3994, 4006, 4008, 4022, 4034, 4044, 4060, 4106, 4112, 4132, 4156, 4174, 4182, 4218, 4240, 4266, 4322, 4386, 4404, 4412, 4420, 4438, 4440, 4444, 4454, 4472, 4518, 4534, 4538, 4584, 4642, 4644, 4646, 4664, 4696, 4700]


# Problem 6

**(a)-(d)** Please check the funciton `quaratic_sieve(n,M,B,trial)` in proj_functions.py

**(e)** The implementation is below. We choose $B = 5000$, $M = 180000$.

In [13]:
B = 5000
M = 180000
n = 10378625636772128629
fact = quadratic_sieve(n,M,B,trials)
if fact:
    print(f'\nFactorization: {n} = {fact} x {n//fact}')

Target: 10378625636772128629 Pseudo Test Result: witness_found: 2, Not prime
Running sieve with M = 180000, B = 5000
Solve left null space using Gaussian on GF(2)...
3 candidates found. Searching...
Factor found: 3628273133
x = 7185082796456316893729664928850186368878847947996128997610492887901703704784739906648707899810061668584258267091109884654428543361430988446161609015940042627099573073095080583993659819347482737250140102769358818727043404815815005174875675292385395252607954495848665396225632975989707533105166869243326467756249079455329650593979387173353665969513373421215289037205136688075901934850148934687105352311409958898790664832925036715694788473415684580885084255341203028092495808934908829773122487766724490730136093994749969729867928859595506531724989055393717690287297021135653918345873081870347832528915580536998773791348103295940489951933011494854229109217797256769618430594988514571268375663235087251718329379888961334454306890464712069284606659499498676881701388713860447425

**(f)** To ensure we get sufficient factor matrix for Gaussian Elimination as well as enough candidate for non-trivial factors. We need to increase both the size of $M$ and $B$. If $B$ is too small, there will no candidates. Since `list_prime` function is inefficient in memory and time and works only for small bound $B$. Such that we constraint $B$ in a reasonable size and mainly focusing on incrasing the size $M$ until first non-trivial factor merged. The parameter above is a trade-off between computation time and number of candidates.